CppCoreGuidelines.md

```diff
@@ -15082,6 +15084,84 @@ Use a function for coroutines.
 Flag a lambda that is a coroutine and has a non-empty capture list.
 
 
+### <a name="Rcoro-locks"></a>CP.52: Do not hold locks or other synchronization primitives across suspension points
+
+##### Reason
+
+This is at best a degradation of performance, and at worst it will cause a deadlock. When a suspension point is reached, such as co_await, execution of the current function stops and other code begins to run. It may be a long period of time before the coroutine resumes. For that entire duration the lock will be held and cannot be acquired by other threads to perform work.
```

Comment:
You could also include the possibility of the function being _never_ resumed. And the possibility of invoking undefined behavior by unlocking some lock in a different thread you described below. Both are stronger arguments than "bad performance", hence I'd put _them_ in a more prominent position and add a note about performance.

Reply:
That's a good argument.  I've flipped it so that deadlock risk is the main reason and poor performance is a note.

Reply:
right, when I explain coroutines (in a different language/framework, but the problems are similar) I just say write as if every line of code is on a different thread. No thread-locals, no locks, no threads.